src/Kestrel.Transport.Sockets/Internal/SocketAwaitableEventArgs.cs

```diff
@@ -5,74 +5,136 @@
+    public class SocketAwaitableEventArgs : SocketAsyncEventArgs, IValueTaskSource<int>
     {
-        private static readonly Action _callbackCompleted = () => { };
+        private static readonly Action<object> _callbackCompleted = _ => { };
+        private static readonly WaitCallback _waitCallback = OnWorkItemCallback;
 
         private readonly PipeScheduler _ioScheduler;
 
-        private Action _callback;
+        private Action<object> _callback;
 
         public SocketAwaitableEventArgs(PipeScheduler ioScheduler)
         {
             _ioScheduler = ioScheduler;
         }
 
-        public SocketAwaitableEventArgs GetAwaiter() => this;
-        public bool IsCompleted => ReferenceEquals(_callback, _callbackCompleted);
+        protected override void OnCompleted(SocketAsyncEventArgs _)
+        {
+            var callback = Interlocked.CompareExchange(ref _callback, _callbackCompleted, null);
+
+            if (callback != null)
+            {
+                Debug.Assert(!ReferenceEquals(callback, _callbackCompleted));
+
+                object state = UserToken;
+                UserToken = null;
 
-        public int GetResult()
+                _ioScheduler.Schedule(callback, state);
+            }
+        }
+
+        public ValueTaskSourceStatus GetStatus(short token)
         {
-            Debug.Assert(ReferenceEquals(_callback, _callbackCompleted));
+            if (ReferenceEquals(_callback, _callbackCompleted))
+            {
+                if (SocketError != SocketError.Success)
+                {
+                    return ValueTaskSourceStatus.Faulted;
+                }
+
+                return ValueTaskSourceStatus.Succeeded;
+            }
+
+            return ValueTaskSourceStatus.Pending;
+        }
+
+        public void OnCompleted(Action<object> continuation, object state, short token, ValueTaskSourceOnCompletedFlags flags)
+        {
+            // We're ignoring ValueTaskSourceOnCompletedFlags and token since this is used in a single place in the code base
+            // so we don't need to handle capturing the SynchronizationContext and ExecutionContext
+            Debug.Assert((flags & ValueTaskSourceOnCompletedFlags.FlowExecutionContext) == 0, "FlowExecutionContext is set.");
+
+            // Use UserToken to carry the continuation state around
+            UserToken = state;
+            Action<object> awaitableState = Interlocked.CompareExchange(ref _callback, continuation, null);
+
+            if (ReferenceEquals(awaitableState, _callbackCompleted))
+            {
+                ThreadPool.UnsafeQueueUserWorkItem(_waitCallback, this);
+            }
+            else if (awaitableState != null)
+            {
+                throw new InvalidOperationException("Multiple continuations registered!");
+            }
 
-            _callback = null;
+        }
 
-            if (SocketError != SocketError.Success)
+        public int GetResult(short token)
```

Comment:
```
The active test run was aborted. Reason: Unhandled Exception: System.InvalidOperationException: Multiple continuations registered!
```

You still need to null out `_callback` here. It's GetCompletedValueTask() where nulling out `_callback` is unnecessary because it will always be null there. Assuming you null it out here that is.

Reply:
😬 ooops